File: public/hello-psi/hello-psi.js
#!/usr/bin/env node
// hello-psi.js — Hello PSI reference implementation.
// Node 18+, zero dependencies. MIT. Free forever. No permission required.
//
// Implements PSI-SEAL/1.0.0 rules R1-R10:
//   R1  RFC 8785 JCS canonicalisation, UTF-8, no BOM
//   R3  field order: schema, schema_digest, sealed_at, subject, hash, merkle
//   R4  lowercase hex digests, 64 chars
//   R5  hash = SHA-256 over the raw octet stream of the subject
//   R6  sealed_at = RFC 3339 UTC, exactly three fractional digits, literal Z
//   R7  subject.name NFC-normalised, subject.size_bytes exact octet length
//   R9  merkle.leaf = SHA-256("PSI1:" + hash)
//   R10 seal_hash = SHA-256(JCS(envelope))
//
// Spec: https://ai-governance-standard.com/.well-known/psi-schema.json
// Verify it yourself:  node hello-psi.js "Hello, PSI."
// ES module, Node 18+.

import { createHash } from "node:crypto";
import { fileURLToPath } from "node:url";

const SCHEMA_ID = "PSI-SEAL/1.0.0";
const VECTOR_SEALED_AT = "2026-01-01T00:00:00.000Z";
const SUBJECT_NAME = "hello-psi-input";
// SHA-256 over the JCS form of the live /.well-known/psi-schema.json.
const SCHEMA_DIGEST = "__SCHEMA_DIGEST__";

export function sha256Hex(input) {
  return createHash("sha256").update(input, typeof input === "string" ? "utf8" : undefined).digest("hex");
}

// RFC 8785 JSON Canonicalization Scheme (subset sufficient for PSI envelopes:
// objects, strings, integers, arrays, booleans, null).
export function jcs(value) {
  if (value === null) return "null";
  const t = typeof value;
  if (t === "boolean") return value ? "true" : "false";
  if (t === "number") {
    if (!Number.isFinite(value)) throw new Error("JCS: non-finite number");
    return Number.isInteger(value) ? String(value) : JSON.stringify(value);
  }
  if (t === "string") return jcsString(value);
  if (Array.isArray(value)) return "[" + value.map(jcs).join(",") + "]";
  if (t === "object") {
    const keys = Object.keys(value)
      .filter((k) => value[k] !== undefined)
      // JCS sorts by UTF-16 code units.
      .sort((a, b) => (a < b ? -1 : a > b ? 1 : 0));
    return "{" + keys.map((k) => jcsString(k) + ":" + jcs(value[k])).join(",") + "}";
  }
  throw new Error("JCS: unsupported type " + t);
}

function jcsString(s) {
  let out = '"';
  for (const ch of s) {
    const c = ch.codePointAt(0);
    if (ch === '"') out += '\\"';
    else if (ch === "\\") out += "\\\\";
    else if (c === 0x08) out += "\\b";
    else if (c === 0x09) out += "\\t";
    else if (c === 0x0a) out += "\\n";
    else if (c === 0x0c) out += "\\f";
    else if (c === 0x0d) out += "\\r";
    else if (c < 0x20) out += "\\u" + c.toString(16).padStart(4, "0");
    else out += ch;
  }
  return out + '"';
}

export function seal(text, sealedAt = VECTOR_SEALED_AT, subjectName = SUBJECT_NAME) {
  const bytes = Buffer.from(text, "utf8");
  const hash = sha256Hex(bytes);
  const leaf = sha256Hex("PSI1:" + hash);
  const envelope = {
    schema: SCHEMA_ID,
    schema_digest: SCHEMA_DIGEST,
    sealed_at: sealedAt,
    subject: { name: subjectName.normalize("NFC"), size_bytes: bytes.length },
    hash,
    merkle: { leaf },
  };
  return { envelope, seal_hash: sha256Hex(jcs(envelope)) };
}

export { SCHEMA_ID, SCHEMA_DIGEST, VECTOR_SEALED_AT };

if (process.argv[1] && fileURLToPath(import.meta.url) === process.argv[1]) {
  const input = process.argv[2] ?? "";
  const result = seal(input);
  console.log(JSON.stringify(result.envelope, null, 2));
  console.log("seal_hash: " + result.seal_hash);
}

File: public/hello-psi/hello_psi.py
#!/usr/bin/env python3
"""hello_psi.py - Hello PSI reference implementation.

Python 3, standard library only (hashlib, json, unicodedata).
MIT. Free forever. No permission required.

Implements PSI-SEAL/1.0.0 rules R1-R10:
  R1  RFC 8785 JCS canonicalisation, UTF-8, no BOM
  R3  field order: schema, schema_digest, sealed_at, subject, hash, merkle
  R4  lowercase hex digests, 64 chars
  R5  hash = SHA-256 over the raw octet stream of the subject
  R6  sealed_at = RFC 3339 UTC, exactly three fractional digits, literal Z
  R7  subject.name NFC-normalised, subject.size_bytes exact octet length
  R9  merkle.leaf = SHA-256("PSI1:" + hash)
  R10 seal_hash = SHA-256(JCS(envelope))

Spec: https://ai-governance-standard.com/.well-known/psi-schema.json
Verify it yourself:  python3 hello_psi.py "Hello, PSI."
"""

import hashlib
import json
import sys
import unicodedata

SCHEMA_ID = "PSI-SEAL/1.0.0"
VECTOR_SEALED_AT = "2026-01-01T00:00:00.000Z"
SUBJECT_NAME = "hello-psi-input"
# SHA-256 over the JCS form of the live /.well-known/psi-schema.json.
SCHEMA_DIGEST = "__SCHEMA_DIGEST__"

_ESCAPES = {
    '"': '\\"',
    "\\": "\\\\",
    "\b": "\\b",
    "\t": "\\t",
    "\n": "\\n",
    "\f": "\\f",
    "\r": "\\r",
}


def sha256_hex(data):
    if isinstance(data, str):
        data = data.encode("utf-8")
    return hashlib.sha256(data).hexdigest()


def _jcs_string(value):
    out = ['"']
    for ch in value:
        if ch in _ESCAPES:
            out.append(_ESCAPES[ch])
        elif ord(ch) < 0x20:
            out.append("\\u%04x" % ord(ch))
        else:
            out.append(ch)
    out.append('"')
    return "".join(out)


def jcs(value):
    """RFC 8785 JSON Canonicalization Scheme (subset used by PSI envelopes)."""
    if value is None:
        return "null"
    if isinstance(value, bool):
        return "true" if value else "false"
    if isinstance(value, int):
        return str(value)
    if isinstance(value, float):
        if value != value or value in (float("inf"), float("-inf")):
            raise ValueError("JCS: non-finite number")
        if value.is_integer():
            return str(int(value))
        return repr(value)
    if isinstance(value, str):
        return _jcs_string(value)
    if isinstance(value, (list, tuple)):
        return "[" + ",".join(jcs(v) for v in value) + "]"
    if isinstance(value, dict):
        # JCS sorts member names by UTF-16 code units.
        keys = sorted(value.keys(), key=lambda k: k.encode("utf-16-be"))
        return "{" + ",".join(_jcs_string(k) + ":" + jcs(value[k]) for k in keys) + "}"
    raise TypeError("JCS: unsupported type %r" % type(value))


def seal(text, sealed_at=VECTOR_SEALED_AT, subject_name=SUBJECT_NAME):
    raw = text.encode("utf-8")
    digest = sha256_hex(raw)
    leaf = sha256_hex("PSI1:" + digest)
    envelope = {
        "schema": SCHEMA_ID,
        "schema_digest": SCHEMA_DIGEST,
        "sealed_at": sealed_at,
        "subject": {
            "name": unicodedata.normalize("NFC", subject_name),
            "size_bytes": len(raw),
        },
        "hash": digest,
        "merkle": {"leaf": leaf},
    }
    return envelope, sha256_hex(jcs(envelope))


if __name__ == "__main__":
    text = sys.argv[1] if len(sys.argv) > 1 else ""
    env, seal_hash = seal(text)
    print(json.dumps(env, indent=2, ensure_ascii=False))
    print("seal_hash: " + seal_hash)
